src/Microsoft.ML.Data/DataLoadSave/TransformWrapper.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Data.DataLoadSave;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Data.IO;
+using Microsoft.ML.Runtime.Model;
+using System.Collections.Generic;
+
+[assembly: LoadableClass(typeof(TransformWrapper), null, typeof(SignatureLoadModel),
+    "Transform wrapper", TransformWrapper.LoaderSignature)]
+
+namespace Microsoft.ML.Runtime.Data
+{
+    // REVIEW: this class is public, as long as the Wrappers.cs in tests still rely on it.
+    // It needs to become internal.
+    public sealed class TransformWrapper : ITransformer, ICanSaveModel
+    {
+        public const string LoaderSignature = "TransformWrapper";
+        private const string TransformDirTemplate = "Step_{0:000}";
+
+        private readonly IHost _host;
+        private readonly IDataView _xf;
+
+        public TransformWrapper(IHostEnvironment env, IDataView xf)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(nameof(TransformWrapper));
+            _host.CheckValue(xf, nameof(xf));
+            _xf = xf;
+        }
+
+        public ISchema GetOutputSchema(ISchema inputSchema)
+        {
+            _host.CheckValue(inputSchema, nameof(inputSchema));
+
+            var dv = new EmptyDataView(_host, inputSchema);
+            var output = ApplyTransformUtils.ApplyAllTransformsToData(_host, _xf, dv);
+            return output.Schema;
+        }
+
+        public void Save(ModelSaveContext ctx)
+        {
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            var dataPipe = _xf;
+            var transforms = new List<IDataTransform>();
+            while (dataPipe is IDataTransform xf)
```

Comment:
>dataPipe is IDataTransform xf [](start = 19, length = 29)

It seems only for linear chain like pipeline. Is not possible to get graph like transform structures?

Reply:
No

---
In reply to: [218249827](https://github.com/dotnet/machinelearning/pull/931#discussion_r218249827) [](ancestors = 218249827)